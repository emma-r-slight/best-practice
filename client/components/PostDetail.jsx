import React from 'react'
import { connect } from 'react-redux'

import { fetchPostById } from '../actions/posts'

class PostDetail extends React.Component {
  componentDidMount () {
    const {post, match, fetchPostById} = this.props
    if (!post && match && match.params) {
      fetchPostById(Number(match.params.id))
    }
  }

  render () {
    const { post = {} } = this.props
    return (
      <li data-testid="post">{post.name}</li>
    )
  }
}

function mapStateToProps (state, ownProps) {
  const id = Number(ownProps.match.params.id)
  const fromState = state.post && state.post.id === id ? state.post : null
  const fromList = state.posts.find(post => post.id === id)
  return {
    post: fromState || fromList
  }
}

const mapDispatchToProps = { fetchPostById }

export default connect(mapStateToProps, mapDispatchToProps)(PostDetail)
