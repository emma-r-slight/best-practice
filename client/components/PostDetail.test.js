import React from 'react'
import { render, screen } from '@testing-library/react'
import { renderWithRedux } from '../testing/utils'
import '@testing-library/jest-dom/extend-expect'
import '@babel/polyfill'

import PostDetail from './PostDetail'
import mockPosts from '../testing/mockPosts'

test('<PostDetail> includes name in <li>', async () => {
  const initialState = {
    posts: mockPosts
  }
  const initialEntries = ['/post/2']
  renderWithRedux(
    <PostDetail match={{params: { id: 2 }}}/>,
    {initialState, initialEntries}
  )
  const post = await screen.findByText('mocked post 2')
  expect(post).toBeInTheDocument()
  expect(post).toMatchSnapshot()
})
