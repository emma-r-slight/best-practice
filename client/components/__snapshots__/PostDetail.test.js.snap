// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostDetail> includes name in <li> 1`] = `
<li
  data-testid="post"
>
  mocked post 2
</li>
`;
